compiler/rustc_lint/src/traits.rs

```diff
@@ -141,3 +141,35 @@ impl<'tcx> LateLintPass<'tcx> for DropTraitConstraints {
         }
     }
 }
+
+declare_lint! {
+    pub DUP_TRAIT_BOUNDS,
+    Warn,
+    "duplicate trait bounds"
+}
+
+declare_lint_pass!(
+    DuplicateTraitBounds => [DUP_TRAIT_BOUNDS]
+);
+
+impl<'tcx> LateLintPass<'tcx> for DuplicateTraitBounds {
+    fn check_ty(&mut self, cx: &LateContext<'_>, ty: &'tcx hir::Ty<'tcx>) {
+        let hir::TyKind::TraitObject(bounds, _lifetime, _syntax) = &ty.kind else {
+            return
+        };
+
+        let mut bounds_vec: Vec<_> = bounds.into_iter().collect();
+
+        for bound in bounds_vec {
+            // LOGIC GOES HERE
+            if bounds_vec.contains(&bound) {
```

Comment:
Two things here:

1) Maybe we can use a `pop` or `remove` to check for duplicates in a loop like this?
2) What's the best way to do the equality check given a [`PolyTraitRef`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.PolyTraitRef.html)? Not a lot of `Eq` implementations on the types here...

Reply:
I believe we can dig all the way in to the corresponding `DefId` of the trait and use that, so we skip the ones that aren't *trait* bounds (like `: 'a` and `<X as Y>::Z: A`) and ignore the others. It'd be nice to also find redundant bounds for those, but we can start with only trait bounds.